src/tests/JIT/Directed/StructABI/StructABI.cs

```diff
@@ -668,6 +668,176 @@ public bool Equals(Nested9 other)
+{
+	long FieldL;
+	Empty FieldE;
+	double FieldD;
+
+	public static LongEmptyDouble Get()
+	{
+		return new LongEmptyDouble { FieldL = 0xcafef00d, FieldD = 3.14159d };
+	}
+
+	public bool Equals(LongEmptyDouble other)
+	{
+		return FieldL == other.FieldL && FieldD.Equals(other.FieldD);
+	}
+}
+
+struct NestedEmpty
+{
+	struct InnerEmpty
+	{
+		Empty e;
+	}
+	InnerEmpty e;
+}
+
+struct NestedEmptyFloatDouble
+{
+	NestedEmpty FieldNE;
+	float FieldF;
+	double FieldD;
+
+	public static NestedEmptyFloatDouble Get()
+	{
+		return new NestedEmptyFloatDouble { FieldF = 3.14159f, FieldD = 3.14159 };
+	}
+
+	public bool Equals(NestedEmptyFloatDouble other)
+	{
+		return FieldF.Equals(other.FieldF) && FieldD.Equals(other.FieldD);
+	}
+}
+
+struct EmptyIntAndFloat
+{
+	struct EmptyInt
+	{
+		public Empty FieldE;
+		public int FieldI;
+	}
+	EmptyInt FieldEI;
+	float FieldF;
+
+	public static EmptyIntAndFloat Get()
+	{
+		return new EmptyIntAndFloat { FieldEI = new EmptyInt { FieldI = 0xcafe }, FieldF = 3.14159f };
+	}
+
+	public bool Equals(EmptyIntAndFloat other)
+	{
+		return FieldEI.FieldI == other.FieldEI.FieldI && FieldF.Equals(other.FieldF);
+	}
+}
+
+struct LongEmptyAndFloat
+{
+	struct LongEmpty
+	{
+		public long FieldL;
+		public Empty FieldE;
+	}
+	LongEmpty FieldLE;
+	float FieldF;
+
+	public static LongEmptyAndFloat Get()
+	{
+		return new LongEmptyAndFloat { FieldLE = new LongEmpty { FieldL = 0xcafef00d }, FieldF = 3.14159f };
+	}
+
+	public bool Equals(LongEmptyAndFloat other)
+	{
+		return FieldLE.FieldL == other.FieldLE.FieldL && FieldF.Equals(other.FieldF);
+	}
+}
+
+[InlineArray(1)]
+struct ArrayOfEmpties
+{
+	Empty e;
+};
+
+struct ArrayOfEmptiesFloatDouble
+{
+	ArrayOfEmpties FieldAoE;
+	float FieldF;
+	double FieldD;
+
+	public static ArrayOfEmptiesFloatDouble Get()
+	{
+		return new ArrayOfEmptiesFloatDouble { FieldF = 3.14159f, FieldD = 3.14159 };
+	}
+
+	public bool Equals(ArrayOfEmptiesFloatDouble other)
+	{
+		return FieldF.Equals(other.FieldF) && FieldD.Equals(other.FieldD);
+	}
+}
+
+struct Eight<T>
+{
+	T e1, e2, e3, e4, e5, e6, e7, e8;
+}
+
+struct FloatEmpty32kInt
+{
+	float FieldF;
+	Eight<Eight<Eight<Eight<Eight<Empty>>>>> FieldEmpty32k;
+	int FieldI;
+
+	public static FloatEmpty32kInt Get()
+	{
+		return new FloatEmpty32kInt { FieldF = 3.14159f, FieldI = 0xcafe };
+	}
+
+	public bool Equals(FloatEmpty32kInt other)
+	{
+		return FieldF.Equals(other.FieldF) && FieldI == other.FieldI;
+	}
+}
+
+[StructLayout(LayoutKind.Sequential, Pack=1)]
+struct PackedEmptyFloatLong
+{
+	Empty FieldE;
+	float FieldF;
+	long FieldL;
+
+	public static PackedEmptyFloatLong Get()
+	{
+		return new PackedEmptyFloatLong { FieldF = 3.14159f, FieldL = 0xcafef00d };
+	}
+
+	public bool Equals(PackedEmptyFloatLong other)
+	{
+		return FieldF.Equals(other.FieldF) && FieldL == other.FieldL;
+	}
+}
+
+[StructLayout(LayoutKind.Explicit, Pack=1)]
+struct ExplicitFloatLong
+{
+	[FieldOffset(1)] float FieldF;
+	[FieldOffset(5)] long FieldL;
+
```

Comment:
For the record, I don't think `Pack` should be necessary. From what I dug in the codebase, it's a known problem:
https://github.com/dotnet/runtime/blob/ce2364c7c3ac0c238a4eea38b36ecc50d67c9964/src/coreclr/vm/jitinterface.cpp#L2102-L2112

IMHO the fix would be something like:
* In CalculateSizeAndFieldOffsets amend calculating alignmentRequirement to:
GCD(min(alignmentRequirement, packingSize), placementInfo->m_offset)
* Amend this condition to take explicit offset into consideration:
https://github.com/dotnet/runtime/blob/ce2364c7c3ac0c238a4eea38b36ecc50d67c9964/src/coreclr/vm/methodtablebuilder.cpp#L4532-L4540

But since this doesn't have much to do with empty structs, I'll leave it for now to keep this PR focused.